include/swift/SIL/SILInstructionWorklist.h

```diff
@@ -24,92 +24,127 @@
+      list = list.drop_back();
+      worklist.insert(instruction);
+    }
+  }
 
   // If instruction is in the worklist, remove it.
-  void remove(SILInstruction *instruction) {
-    auto iterator = worklistMap.find(instruction);
-    if (iterator == worklistMap.end())
-      return; // Not in worklist.
-
-    // Don't bother moving everything down, just null out the slot. We will
-    // check before we process any instruction if it is null.
-    worklist[iterator->second] = nullptr;
-    worklistMap.erase(iterator);
-  }
+  void erase(SILInstruction *instruction) { worklist.erase(instruction); }
 
   /// Remove the top element from the worklist.
-  SILInstruction *removeOne() {
-    SILInstruction *instruction = worklist.pop_back_val();
-    worklistMap.erase(instruction);
-    return instruction;
+  SILInstruction *pop_back_val() {
+    return worklist.pop_back_val().getValueOr(nullptr);
   }
 
   /// When an instruction has been simplified, add all of its users to the
   /// worklist, since additional simplifications of its users may have been
   /// exposed.
-  void addUsersToWorklist(ValueBase *instruction) {
-    for (auto *use : instruction->getUses())
-      add(use->getUser());
-  }
-
   void addUsersToWorklist(SILValue value) {
-    for (auto *use : value->getUses())
+    for (auto *use : value->getUses()) {
       add(use->getUser());
+    }
   }
 
   /// When an instruction has been simplified, add all of its users to the
   /// worklist, since additional simplifications of its users may have been
   /// exposed.
   void addUsersOfAllResultsToWorklist(SILInstruction *instruction) {
     for (auto result : instruction->getResults()) {
-      addUsersToWorklist(result);
+      addUsersToWorklist(&*result);
     }
   }
 
-  /// Check that the worklist is empty and nuke the backing store for the map if
-  /// it is large.
-  void zap() {
-    assert(worklistMap.empty() && "Worklist empty, but the map is not?");
+  /// Check that the worklist is empty and nuke the backing store if it is
+  /// large.
+  void resetChecked() {
+    assert(worklist.empty() && "Vector empty, but the map is not?");
 
-    // Do an explicit clear, this shrinks the map if needed.
-    worklistMap.clear();
+    // Do an explicit clear, shrinking the storage if needed.
+    worklist.clear();
   }
 };
 
+// TODO: This name is somewhat unpleasant.  Once the type is templated over its
+//       storage and renamed BlottableWorklist, this type will be
+//       SmallBlottableWorklist which is more reasonable.
+template <unsigned N>
+class SmallSILInstructionWorklist final
+    : public SILInstructionWorklist<
+          llvm::SmallVector<Optional<SILInstruction *>, N>,
+          llvm::DenseMap<SILInstruction *, unsigned>> {
+public:
+  using VectorT = llvm::SmallVector<Optional<SILInstruction *>, N>;
+  using MapT = llvm::DenseMap<SILInstruction *, unsigned>;
+  SmallSILInstructionWorklist(StringLiteral loggingName = "InstructionWorklist")
```

Comment:
Just an FYI, I think I mislead you here. Sorry! I was checking the code in LLVM and it says that StringLiteral is only supposed to be used for global char * strings. I would just follow what Andy said and use the const char *.

Reply:
Fixed.

Reply:
@gottesmm @nate-chandler 
Cool. Maybe I should clarify for future reference:
- `const char *` is "code smell" because it's not best practice C++
- Hiding the smell with `StringRef` is deceptive because, if the logic is "this is only correct when used with a `const char *`" literal, then that should be directly communicated by using `const char *`
- Doing the non-stinky thing and storing `std::string` is overkill and unnecessarily costly. 

So, for a field whose only purpose is to allow reuse over a few internal literal strings, `const char *` is the perfect amount of smelliness that nicely alerts the programmer that we're caching a pointer to an internal string literal.

Reply:
Noted!  From here on "the perfect amount of smelliness" will be used.  😄